src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs

```diff
@@ -0,0 +1,353 @@
+﻿using System;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public class AttributeTests_Tuples : WellKnownAttributesTestBase
+    {
+        private static readonly string s_tuplesTestSource = @"
+public class Base0 { }
+public class Base1<T> { }
+public class Base2<T, U> { }
+
+public class Outer<T> : Base1<(int key, int val)>
+{
+    public class Inner<U, V> : Base2<(int key2, int val2), V>
+    {
+        public class InnerInner<W> : Base1<(int key3, int val3)> { }
+    }
+}
+
+public class Derived<T> : Outer<(int e1, (int e2, int e3) e4)>.Inner<
+    Outer<(int e5, int e6)>.Inner<(int e7, int e8)[], (int e9, int e10)>.InnerInner<int>[],
+    ((int e11, int e12) e13, int e14)>
+    .InnerInner<((int e15, int e16) e17, (int e18, Base1<(int e19, int e20)> e21) e22)>
+{
+    public static (int e1, int e2) Field1;
+    public static (int e1, int e2) Field2;
+    public static Base1<(int e1, (int e2, int e3) e4)> Field3;
+
+    public static (int e1, int e2) Method1() => (0, 0);
+    public static void Method2((int e1, int e2) x) { }
+    public static (int e1, int e2) Method3((int e3, int e4) x) => (0, 0);
+    public static (int e1, int e2) Method4(ref (int e3, int e4) x) => x;
+
+    public static (int e1, int e2) Prop1 => (0, 0);
+    public static (int e1, int e2) Prop2 { get; set; }
```

Comment:
need scenarios for:
-  not tuplefied valuetuple like      'ValueTuple<t1, t2>  foo()'
-  tuples with no named elements   like   '(int, int) foo()'
- arrays/generics involving tuples  - like    ' List<(int a, int b)?> foo()'
- various combinations of the above trying to force the decoder to misread names.   - cases like   `(ValueTuple<int, (int a, int b)> c, int d, (int, int) e, int f) Foo() .... ` etc